lib/rust/flexer/src/lib.rs

```diff
@@ -14,26 +15,220 @@
-// === Flexer ====
-// ===============
+
+// ==============
+// === Flexer ===
+// ==============
 
 /// The flexer is an engine for generating lexers.
 ///
 /// Akin to flex and other lexer generators, it is given a definition as a series of rules from
 /// which it then generates code for a highly optimised lexer implemented on top of a
 /// [DFA](https://en.wikipedia.org/wiki/Deterministic_finite_automaton).
-pub trait Flexer {
+#[derive(Clone,Debug)]
+pub struct Flexer<Definition,Output,Reader> where Reader:LazyReader {
+    /// The stack of states that are active during lexer execution.
+    pub state_stack: Vec<usize>,
+    /// A reader for the input.
+    pub reader: Reader,
+    /// The current match of the lexer.
+    pub current_match: String,
+    /// The result of the current stage of the DFA.
+    pub status: FlexerStageStatus,
+    /// The tokens that have been lexed.
+    pub tokens: Vec<Output>,
+    /// The definition of the lexer.
+    definition: Definition
+}
+
+impl<Definition,Output,Reader> Deref for Flexer<Definition,Output,Reader>
+where Reader:LazyReader {
+    type Target = Definition;
+    fn deref(&self) -> &Self::Target {
+        &self.definition
+    }
+}
+
+impl<Definition,Output,Reader> DerefMut for Flexer<Definition,Output,Reader>
+where Reader:LazyReader {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.definition
+    }
+}
+
+impl<Definition,Output,Reader> Flexer<Definition,Output,Reader>
+where Definition: FlexerState<Reader>,Reader:LazyReader {
+    /// Creates a new lexer instance.
+    ///
+    /// Please note that the `reader` argument is currently hard-coded for testing purposes. This is
+    /// not the intention for the eventual design.
+    pub fn new(mut reader:Reader) -> Flexer<Definition,Output,Reader> {
+        let mut state_stack = Vec::new();
+        state_stack.reserve(1024);
+        let current_match = String::from("");
+        let status = FlexerStageStatus::Initial;
+        let mut tokens = Vec::new();
+        tokens.reserve(1024);
+        let definition = Definition::new(&mut reader);
+        let initial_state_id = definition.initial_state().id;
+        state_stack.push(initial_state_id);
+
+        Flexer {state_stack,reader,current_match,status,tokens,definition}
+    }
+}
+
+/// This block is things that are part of the lexer's interface and functionality.
+impl<Definition,Reader,Output> Flexer<Definition,Output,Reader>
+where Definition:FlexerState<Reader>,Reader:LazyReader,Output:Clone {
+    /// Gets the lexer result.
+    pub fn get_result(&mut self) -> Option<Vec<Output>> {
+        Some(self.tokens.clone())
+    }
+
+    /// Gets the lexer's root state.
+    pub fn root_state(&self) -> &Group {
+        &self.definition.initial_state()
+    }
+
+    /// Gets the state that the lexer is currently in.
+    pub fn current_state(&self) -> usize {
+        *self.state_stack.last().expect("There should always be one state on the stack.")
```

Comment:
using `usize` - this should be newtype if its exposed to public API. Applies ofc to other functions like `begin_state`